sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadViaSDF.java

```diff
@@ -0,0 +1,742 @@
+
+      KafkaLatestOffsetEstimator(
+          Consumer<byte[], byte[]> offsetConsumer, TopicPartition topicPartition) {
+        this.offsetConsumer = offsetConsumer;
+        this.topicPartition = topicPartition;
+        this.consumerSpEL = new ConsumerSpEL();
+        this.consumerSpEL.evaluateAssign(
+            this.offsetConsumer, ImmutableList.of(this.topicPartition));
+      }
+
+      @Override
+      protected void finalize() {
+        try {
+          Closeables.close(offsetConsumer, true);
+        } catch (Exception anyException) {
+          LOG.warn("Failed to close offset consumer for {}", topicPartition);
+        }
+      }
+
+      @Override
+      public long estimate() {
+        consumerSpEL.evaluateSeek2End(offsetConsumer, topicPartition);
+        return offsetConsumer.position(topicPartition);
+      }
+    }
+
+    @GetInitialRestriction
+    public OffsetRange initialRestriction(@Element Row kafkaSourceDescription) {
+      Map<String, Object> updatedConsumerConfig =
+          overrideBootstrapServersConfig(consumerConfig, kafkaSourceDescription);
+      try (Consumer<byte[], byte[]> offsetConsumer =
+          consumerFactoryFn.apply(
+              KafkaIOUtils.getOffsetConsumerConfig(
+                  "initialOffset", offsetConsumerConfig, updatedConsumerConfig))) {
+        TopicPartition topicPartition =
+            new TopicPartition(
+                kafkaSourceDescription.getString(KafkaSourceDescriptionSchemas.TOPIC),
+                kafkaSourceDescription.getInt32(KafkaSourceDescriptionSchemas.PARTITION));
+        consumerSpEL.evaluateAssign(offsetConsumer, ImmutableList.of(topicPartition));
+        long startOffset;
+        if (kafkaSourceDescription.getInt64(KafkaSourceDescriptionSchemas.START_READ_OFFSET)
+            != null) {
+          startOffset =
+              kafkaSourceDescription.getInt64(KafkaSourceDescriptionSchemas.START_READ_OFFSET);
+        } else if (kafkaSourceDescription.getInt64(KafkaSourceDescriptionSchemas.START_READ_TIME)
+            != null) {
+          startOffset =
+              consumerSpEL.offsetForTime(
+                  offsetConsumer,
+                  topicPartition,
+                  Instant.ofEpochMilli(kafkaSourceDescription.getInt64("start_read_time")));
+        } else {
+          startOffset = offsetConsumer.position(topicPartition);
+        }
+        return new OffsetRange(startOffset, Long.MAX_VALUE);
+      }
+    }
+
+    @GetInitialWatermarkEstimatorState
+    public Instant getInitialWatermarkEstimatorState(@Timestamp Instant currentElementTimestamp) {
+      return currentElementTimestamp;
+    }
+
+    @NewWatermarkEstimator
+    public MonotonicallyIncreasing newWatermarkEstimator(
+        @WatermarkEstimatorState Instant watermarkEstimatorState) {
+      return new MonotonicallyIncreasing(watermarkEstimatorState);
+    }
+
+    @GetSize
+    public double getSize(@Element Row kafkaSourceDescription, @Restriction OffsetRange offsetRange)
```

Comment:
Why not invoke `initialRestriction` and pass in the necessary arguments instead of duplicating the setup work.

Reply:
`initialRestriction` still gives the infinite range. I should invoke `restrictionTracker().getProgress()`

Reply:
My bad, totally meant to say `restrictionTracker()`